sw/host/opentitanlib/src/proxy/socket_server.rs

```diff
@@ -0,0 +1,298 @@
+        poll.registry()
+            .register(&mut socket, socket_token, Interest::READABLE)?;
+        // Create a `Signals` instance that will catch given set of signals for us.
+        let signals: SignalSet = Signal::Terminate | Signal::Interrupt;
+        let mut signals = Signals::new(signals)?;
+        // And register it with our `Poll` instance.
+        let signal_token = get_next_token();
+        poll.registry()
+            .register(&mut signals, signal_token, Interest::READABLE)?;
+        Ok(Self {
+            command_handler,
+            poll,
+            socket,
+            socket_token,
+            signals,
+            signal_token,
+            connection_map: HashMap::new(),
+            exit_requested: false,
+            phantom: std::marker::PhantomData,
+        })
+    }
+
+    pub fn run_loop(&mut self) -> Result<()> {
+        let mut events = Events::with_capacity(1024);
+        while !self.exit_requested {
+            self.poll.poll(&mut events, None)?;
+            for event in events.iter() {
+                if event.token() == self.socket_token {
+                    self.process_new_connection()?;
+                } else if event.token() == self.signal_token {
+                    self.process_signals()?;
+                } else {
+                    match self.process_connection(event) {
+                        Ok(false) => (),
+                        Ok(true) => self.shutdown_connection(event)?,
+                        Err(e) => {
+                            log::warn!(
+                                "Connection {:#X} error: {}",
+                                event.token().0,
+                                e,
+                            );
+                            self.shutdown_connection(event)?;
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Accept new socket connections, creating new Connection objects.
+    fn process_new_connection(&mut self) -> Result<()> {
+        loop {
+            match self.socket.accept() {
+                Ok((mut conn_socket, _addres)) => {
+                    let token = get_next_token();
+                    log::info!("New connection id:{:#X}", token.0);
+                    match self.connection_map.entry(token) {
+                        Vacant(entry) => {
+                            self.poll.registry().register(
+                                &mut conn_socket,
+                                token,
+                                Interest::READABLE | Interest::WRITABLE,
+                            )?;
+                            entry.insert(Connection::new(conn_socket));
+                        }
+                        Occupied(_) => {
+                            panic!("JsonSocketServer error: token colision");
+                        }
+                    };
+                }
+                Err(err) => {
```

Comment:
I like to split these into match arms with guards:
```rust
Err(err) if err.kind() == ErrorKind::WouldBlock =>
Err(err) =>
```

Reply:
Done